include/oneapi/dpl/internal/scan_by_segment_impl.h

```diff
@@ -0,0 +1,506 @@
+
+                    // 1b. Perform a work group scan to find the carry in value to apply to each item.
+                    auto __closest_seg_id = sycl::inclusive_scan_over_group(__group, __max_end, sycl::maximum<>());
+
+                    __val_type __carry_in =
+                        wg_segmented_scan(__item, __loc_acc, __local_id, __local_id - __closest_seg_id, __accumulator,
+                                          __binary_op, __wgroup_size);
+
+                    // 1c. Update partial scan values and write to global memory.
+                    if (__local_id != 0 && __start < __n && __keys[__start] == __keys[__start - 1])
+                    {
+                        for (int32_t __i = __start; __i < __end; ++__i)
+                        {
+                            __out_values[__i] = __binary_op(__carry_in, __out_values[__i]);
+
+                            if (__i == __n - 1 || __keys[__i] != __keys[__i + 1])
+                                break;
+                        }
+                    }
+                    
+                    // 1d. Store partials and presence of segment flags for use in phase 2
+                    if (__local_id == __wgroup_size - 1) // last work item writes the group's carry out
+                    {
+                        auto __group_id = __group.get_group_id(0);
+
+                        // If no segment ends in the item, the aggregates from previous work groups must be applied.
+                        if (__max_end == 0)
+                            __partials_acc[__group_id] =
+                                __binary_op(__carry_in, __accumulator); // inclusive with last element
+
+                        else
+                            __partials_acc[__group_id] = __accumulator;
+
+                        bool __seg_present = static_cast<bool>(__closest_seg_id);
+                        __seg_ends_acc[__group_id] = __seg_present;
+                    }
+                });
+        });
+
+        // 2. Apply work group carry outs, calculate output indices, and load results into correct indices.
+        __exec.queue()
+            .submit([&](sycl::handler& __cgh) {
+                oneapi::dpl::__ranges::__require_access(__cgh, __keys, __out_values);
+
+                auto __partials_acc = __partials.template get_access<sycl::access_mode::read>(__cgh);
+                auto __seg_ends_acc = __seg_ends.template get_access<sycl::access_mode::read>(__cgh);
+
+                __cgh.depends_on(__wg_incl_scan);
+
+                __cgh.parallel_for(
+                    sycl::nd_range<1>{__n_groups * __wgroup_size, __wgroup_size}, [=](sycl::nd_item<1> __item) {
+                        auto __group = __item.get_group();
+                        auto __group_id = __group.get_group_id(0);
+                        auto __global_id = __item.get_global_id();
+                        auto __local_id = __item.get_local_id();
+
+                        int __wg_agg_idx = __group_id - 1;
```

Comment:
You assign here std::size_t to int. Is it correct?

Reply:
Yes, in this case this correct since the group's id may be 0, so we must use a signed type.